### PR TITLE
Add include_data :via_include_parameter

### PR DESCRIPTION
**Pull request by richmolj**
@NullVoxPopuli @bf4, taking a final stab at simplifying the code from https://github.com/rails-api/active_model_serializers/pull/1720. I created a separate PR since the original solves more issues and I didn't want to blow away the tests if you wanted to use them. Feel free to just close this PR if you still want to go in another direction, don't want this to seem pushy 😄 I believe this is the simplest possible solution, so maybe helpful to understand the problem if nothing else.

This now adds only one new variable name, `include_slice`. This is the slice of the include directive hash being  processed by the current serializer. All the changes to `lib/active_model/serializer/associations.rb` and `lib/active_model_serializers/adapter/json_api.rb` are just passing that slice around so we can access it when we finally decide whether or not to have the `data` key in the response. The changes to `lib/active_model/serializer/reflection.rb` access that variable to make the decision.

Why do we need the current slice instead of the whole directive? Because multiple entities can have the same relationship, e.g. `/blogs/?include=posts.tags,tags` should include both blog tags and post tags, but `/blogs/?include=posts.tags` should only include post tags

Users now opt-in to this pattern **using the current DSL**. The current DSL conditionally includes the `data` key in the response via `include_data true/false`:

```ruby
class PostSerializer < ActiveModel::Serializer
  has_many :comments do
    include_data false
  end
end
```

We now add another option `:via_include_parameter`. This means we only include the `data` key when the relationship is being sideloaded:

```ruby
class PostSerializer < ActiveModel::Serializer
  has_many :comments do
    include_data :via_include_parameter
  end
end
```

### CODE REVIEW

File: lib/active_model_serializers/adapter/json_api.rb [hunk 1]
```diff
@@ -235,17 +235,17 @@ def resource_objects_for(serializers)
...
         @resource_identifiers = Set.new
-        serializers.each { |serializer| process_resource(serializer, true) }
+        serializers.each { |serializer| process_resource(serializer, true, @include_directive) }
         serializers.each { |serializer| process_relationships(serializer, @include_directive) }
 
         [@primary, @included]
       end
 
-      def process_resource(serializer, primary)
+      def process_resource(serializer, primary, include_slice = {})
```
Comment (bf4) [inline lib/active_model_serializers/adapter/json_api.rb]: directive now aka slice

File: lib/active_model_serializers/adapter/json_api.rb [hunk 2]
```diff
@@ -432,12 +432,12 @@ def resource_object_for(serializer)
       #     id: 'required-id',
       #     meta: meta
       #   }.reject! {|_,v| v.nil? }
-      def relationships_for(serializer, requested_associations)
+      def relationships_for(serializer, requested_associations, include_slice)
         include_directive = JSONAPI::IncludeDirective.new(
           requested_associations,
```
Comment (bf4) [inline lib/active_model_serializers/adapter/json_api.rb]: now have slice and directive

**Comment by bf4** [comment]
I feel like we shouldn't need to introduce the include_slice directive (i.e have two include directives) for this to work.  Otherwise, the public interface change is good.

File: test/adapter/json_api/include_param_test.rb [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+          def setup
+            IncludeParamAuthorSerializer.comment_loader = Class.new(CustomCommentLoader).new
+            @tag = Tag.new(id: 1337, name: 'mytag')
+            @author = IncludeParamAuthor.new(
+              id: 1337,
+              tags: [@tag]
+            )
+          end
+
+          def test_relationship_not_loaded_when_not_included
```
Comment (bf4) [inline test/adapter/json_api/include_param_test.rb]: I think this test is the central feature in this PR, at least for the JSONAPI adapter

**Comment by richmolj** [comment]
> I feel like we shouldn't need to introduce the include_slice directive (i.e have two include directives) for this to work. Otherwise, the public interface change is good.

I feel you, but not sure how to avoid it. If you have an `include_directive` like `{ posts: { comments: :author }, reviews: { comments: {} }`, and you only ever know about the entire directive and not the slice currently being processed, how would `CommentSerializer` know if `author` data should be populated? It's conditional depending on the nesting level.

**Comment by richmolj** [comment]
> It should be modifying the include and fields params as each serializer calls the next, no?

@bf4 This may be a matter of phrasing, but to be clear - since `fields` and `include` live on the **adapter**, the adapter recursively iterates through associations (one serializer doesn't call the next - I wish they did, the issue would likely be avoided). During iteration, since we're only dealing with a slice of the overall directive, I renamed the parameter named `include_slice` to make this more obvious.

I feel like the current PR is already doing what you are suggesting, so let me try to rephrase - I think the confusion is with the current codebase, not this PR. `include_directive` means **3 different things** in the current codebase:

* The `@include_directive` - the overall directive passed to `render`
* The `include_directive` in parameters that I now refer to as `include_slice`. This is the slice of the `@include_directive` we're currently iterating over.
* The `include_directive` in `#relationships_for` is a totally separate concept, derived from the `fields` param and used to iterate over associations.

IOW, you could actually refer to it as `include_slice` or `include_directive` **everywhere**...except the [current implementation of #relationships_for](https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model_serializers/adapter/json_api.rb#L435-L450) screws us up. This introduces [another variable named 'include_directive'](https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model_serializers/adapter/json_api.rb#L436-L439) that is **not the include directive we've been dealing with this whole time**. It's a separate directive [derived from the 'fields' parameter](https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model_serializers/adapter/json_api.rb#L306-L307). It uses this new directive [to iterate over the associations](https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model_serializers/adapter/json_api.rb#L440) - and we can't pass only the current slice (you'll get a ton of test failures as we often need to process relationships even when `include` is not present). But we need that current slice down the line in `reflection.rb`, which is why we end up referring to both `include_directive` and `include_slice` in the same method and confusion ensues.

Hope this helps - as always, your time and effort here are much appreciated.

**Comment by NullVoxPopuli** [comment]
Some thoughts:

I don't think `fields` should be recursive, unless we added the same notation that we do for include (`'field1,field2.subfield1`).

What if the API was this:

```ruby
# current behavior
has_many :comments


has_many :comments do
  # must explicitly be included
  # or maybe something related to the resource identifier object?
  # http://jsonapi.org/format/#document-resource-identifier-objects
  skip_unless_included
end
```
I feel like `include_data` is a little ambiguous, as what we want to avoid is the Resource Identifier object isn't it?



I don't like the additional `include_` variable.

**Comment by bf4** [comment]
Hmm, fields_directive...

B mobile phone

**Comment by NullVoxPopuli** [comment]
> Hmm, fields_directive...

at least initially, we could just have fields span attributes and relationships, flatly. No need for a fields directive, I don't think. at least.. not yet.

**Comment by richmolj** [comment]
@NullVoxPopuli correct me if I'm wrong, but `fields` is not currently recursive. I'm actually not sure how/why it would ever be - JSONAPI [fields](http://jsonapi.org/format/#fetching-sparse-fieldsets) is flat and namespaced, e.g. ` /articles?include=author&fields[articles]=title,body&fields[people]=name`. Something like `field1.subfield1` would deviate from the spec AFAIK. [#process_relationships](https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model_serializers/adapter/json_api.rb#L258-L261) is the source of recursion here.

> skip_unless_included

:+1: to `include_data` being ambiguous, but keep in mind the current behavior of this gem uses `include_data false` independent of the `?include` parameter. So replacing with `skip_unless_included` probably doesn't match well for the current use cases you guys are supporting.

> I don't like the additional include_ variable.

I think that is something everyone, including me, agrees on :joy: My thing is: what's the alternative? I can't see how you avoid it without fundamentally changing how `adapter/json_api.rb` works (avoiding the recursion and switching to a serializers-call-serializers pattern).

**Comment by NullVoxPopuli** [comment]
> JSONAPI fields is flat and namespaced

I think we're on the same page here.

> So replacing with skip_unless_included probably doesn't match well for the current use cases you guys are supporting.

how do you mean?

> avoiding the recursion and switching to a serializers-call-serializers pattern

ah, I see. that is obnoxious then.

**Comment by bf4** [comment]
`skip_unless_included` should be default for what jsonapi passes into serializer 'attributes'

**Comment by bf4** [comment]
i.e.

```ruby
# https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model/serializer.rb
    def serializable_hash(adapter_options = nil, options = {}, adapter_instance = self.class.serialization_adapter_instance)
      resource = fetch_attributes(options[:fields], cached_attributes, adapter_instance)
# https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model/serializer/attributes.rb
        def attributes(requested_attrs = nil, reload = false)
 next if attr.excluded?(self)
            next unless requested_attrs.nil? || requested_attrs.include?(key)
```

**Comment by richmolj** [comment]
@NullVoxPopuli regarding `skip_unless_included` and the current behavior of the gem: Assuming people are using `include_data false` right now with the current gem, they are excluding the `data` key **independent of the ?include param**. So renaming this option something specific to the `?include` param (`skip_unless_included`) seems off to me. That said, I agree with your general thought and for my personal use of AMS I'd be :+1:

**Comment by NullVoxPopuli** [comment]
> skip_unless_included should be default for what jsonapi passes into serializer 'attributes'

@bf4, do you mean we _shouldn't_ include the id/type for relationships (by default) (no options passed)

> Assuming people are using include_data false right now with the current gem, they are excluding the data key

@richmolj  I actually didn't know that was a thing. ha

**Comment by richmolj** [comment]
@bf4 while I agree `skip_unless_included` should be a default, I am not sure 'attributes' is the place for it, for two reasons:

One - I *think* the payload output would be something like:

```
{
  relationships: {
    author: {
      data: {
        id: 1, type: 'people', attributes: {}
      }
    }
  }
}
```

Which means the resource identifiers would be included, a DB hit is still happening, and we haven't solved the issue.

That said, even if we got the payload to look like this:

```
{
  relationships: {
    author: {
      data: {}
    }
  }
}
```

It's still a no-go. This is telling clients "this object does not have an author", versus "this object may or may not have an author, we're just not including author information in the response".

**Comment by bradens** [comment]
I need this so bad, is there anything I can do to help get something to solve this problem released?

**Comment by richmolj** [comment]
@bf4 - per our discussion yesterday:

* Renamed `via_include_parameter` to `if_sideloaded`, so it now reads `include_data :if_sideloaded`.
* Added `ActiveModel::Serializer.config.include_data_default` configuration option.

Though we both agree this should be the default, I didn't change the current default out of backwards compatibility concerns. If you're OK with it I'm OK with it, though.

File: lib/active_model/serializer/concerns/configuration.rb [hunk 4]
```diff
@@ -29,6 +29,7 @@ def config.array_serializer
         # Make JSON API top-level jsonapi member opt-in
         # ref: http://jsonapi.org/format/#document-top-level
         config.jsonapi_include_toplevel_object = false
+        config.include_data_default = true
```
Comment (bf4) [inline lib/active_model/serializer/concerns/configuration.rb]: jsonapi adapter should default to false, right?

File: lib/active_model/serializer/reflection.rb [hunk 5]
```diff
@@ -69,19 +69,17 @@ def include_data(value = true)
...
-          if block_value != :nil
-            block_value
-          elsif @_include_data
-            serializer.read_attribute_for_serialization(name)
-          end
-        else
+        block_value = instance_exec(serializer, &block) if block
+        return unless include_data?(include_slice)
+
+        if [nil, :nil].include?(block_value)
```
Comment (bf4) [inline lib/active_model/serializer/reflection.rb]: nil is valid iirc

File: lib/active_model/serializer/reflection.rb [hunk 6]
```diff
@@ -137,6 +135,14 @@ def build_association(parent_serializer, parent_serializer_options)
 
       private
 
+      def include_data?(include_slice)
+        if @_include_data == :if_sideloaded
+          include_slice.key?(name)
+        else
+          @_include_data
+        end
```
Comment (bf4) [inline lib/active_model/serializer/reflection.rb]: 👍

File: lib/active_model_serializers/adapter/json_api.rb [hunk 7]
```diff
@@ -304,7 +304,7 @@ def resource_object_for(serializer)
         end
 
         requested_associations = fieldset.fields_for(resource_object[:type]) || '*'
```
Comment (bf4) [inline lib/active_model_serializers/adapter/json_api.rb]: urg, `*`

File: lib/active_model/serializer/concerns/configuration.rb (hunk 4 above)
Comment (richmolj) [inline lib/active_model/serializer/concerns/configuration.rb]: current default is true. per the PR description, I vote we change that default but up to you guys

File: lib/active_model/serializer/concerns/associations.rb [hunk 8]
```diff
@@ -83,15 +83,17 @@ def associate(reflection)
...
-      def associations(include_directive = ActiveModelSerializers.default_include_directive)
+      def associations(include_directive = ActiveModelSerializers.default_include_directive, include_slice = nil)
+        include_slice ||= include_directive
         return unless object
 
         Enumerator.new do |y|
           self.class._reflections.values.each do |reflection|
             next if reflection.excluded?(self)
             key = reflection.options.fetch(:key, reflection.name)
             next unless include_directive.key?(key)
```
Comment (bf4) [inline lib/active_model/serializer/concerns/associations.rb]: in this context, when would `include_directive.key?(key)` not be equal to `include_slice.key?(key)`?

File: lib/active_model/serializer/reflection.rb (hunk 5 above)
Comment (richmolj) [inline lib/active_model/serializer/reflection.rb]: great catch 👍 , made the change

File: lib/active_model/serializer/concerns/associations.rb (hunk 8 above)
Comment (richmolj) [inline lib/active_model/serializer/concerns/associations.rb]: This is one of the gross aspects of the current adapter but

* we get the [fields](https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model_serializers/adapter/json_api.rb#L306), defaulted to '*'.

* they end up in [relationships_for](https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model_serializers/adapter/json_api.rb#L435)

* that calls [#associations](https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model_serializers/adapter/json_api.rb#L440) leading us to this code.

That's opposed to `include_slice`, which is always the slice of the *real* include directive passed to `render`.

In other words, the current code has two 'include directives'. One real one that is what you think it is, and a separate one that is derived from a relationship's `fields`

Comment (richmolj) [inline lib/active_model/serializer/concerns/associations.rb]: So in this case, for instance, let's say your include directive was:

```ruby
{ posts: { comments: {} } }
```

When you are iterating the relations and you hit the `comments` portion, the `include_directive` (the current code) is actually '*', assuming no sparse fieldsets requested. But the `include_slice` would be {}.

**Comment by kevintyll** [comment]
Is there any concern for ensuring that the relationship object contains at least of of `data`, `links`, or `meta` elements?
Per the spec, one is required, but if the association is set up with no link, or meta defined, and `include_data :if_sideloaded`, then the result will not be compliant?

Looking forward to this feature BTW.  I just implemented my own monkey patch to do exactly this.  This solution is better.

**Comment by richmolj** [comment]
@kevintyll - definitely agree. If the relation is not influded you should see an empty meta key, eg 'comments: { meta: {} }'. I'm on mobile right now but I should add a test asserting that is the case.

File: lib/active_model_serializers/adapter/json_api.rb (hunk 7 above)
Comment (NullVoxPopuli) [inline lib/active_model_serializers/adapter/json_api.rb]: :-(

**Comment by bf4** [comment]
@richmolj issue title /description should also be updated, I think, no?

**Comment by richmolj** [comment]
@bf4 agreed, will close this and submit a new one tonight.

**Comment by richmolj** [comment]
Moved to https://github.com/rails-api/active_model_serializers/pull/1931
